csharp/src/Ice/Reference.cs

```diff
@@ -1820,5 +1518,292 @@ internal void Write(OutputStream ostr)
+                // is established or the request sent and idempotent
+                Debug.Assert(response != null || exception != null);
+                RetryPolicy retryPolicy =
+                    response?.GetRetryPolicy(this) ?? exception!.GetRetryPolicy(request.IsIdempotent, sent);
+
+                // With the retry-policy OtherReplica we add the current connector to the list of excluded
+                // connectors and remove if from the list of connectors, this prevents the connector to be
+                // tried again during the current retry sequence.
+                if (retryPolicy == RetryPolicy.OtherReplica)
+                {
+                    if ((connectors?[nextConnector] ?? connection?.Connector) is Connector connector)
+                    {
+                        excludedConnectors ??= new();
+                        excludedConnectors.Add(connector);
+                    }
+                    connectors?.RemoveAt(nextConnector);
+                }
+
+                if (connectors != null)
+                {
+                    if (connection == null && retryPolicy != RetryPolicy.OtherReplica)
+                    {
+                        // If connection establishment failed and the connector was not excluded, try the next
+                        // connector
+                        nextConnector = ++nextConnector % connectors.Count;
+                    }
+
+                    if (connectors.Count == 0 || nextConnector == 0)
+                    {
+                        // If the connector set is empty because all connectors has been excluded, or
+                        // nextConnector == 0, it indicates that we already tried all the connectors.
+                        if (cached)
+                        {
+                            // If the connectors were computed from cached endpoints, we clear the connectors to
+                            // trigger a new endpoint lookup.
+                            connectors = null;
+                            nextConnector = 0;
+                        }
+                        else
+                        {
+                            // Otherwise we set triedAllConnectors to true to ensure further connection establishment
+                            // failures will now count as attempts (to prevent indefinitely looping if connection
+                            // establishment failure results in a retryable exception).
+                            triedAllConnectors = true;
+                        }
+                    }
+                }
+
+                // Check if we can retry, we cannot retry if we have consumed all attempts, the current retry
+                // policy doesn't allow retries, the request was already released, there are no more connectors
+                // or a fixed reference receives an exception with OtherReplica retry.
+                if (attempt == Communicator.MaxAttempts ||
```

Comment:
I find this:
```
while (true)
{
    // bunch of code

    // check if last attempt or other condition and return

    // bunch of other code
}
```

not very elegant or readable.

Reply:
So what about something like:
```
IncomingResponseFrame? response = null;
Exception? exception = null;
bool canRetry;
do
{
    // bunch of code

    // check if last attempt or other condition
    canRetry = ...
    if (canRetry)
    {
        // bunch of other code
    }
}
while (canRetry);

return response ?? throw exception;
```

Reply:
Yes, and maybe simply retry or tryAgain for the variable?

```
bool tryAgain = false;
do
{

}
while (tryAgain);
```

Reply:
added this `do{....} while(tryAgain);`

Reply:
To me the `while(true)` loop was simpler than this loop with a boolean ;-)

Reply:
I don't have a strong preference here